# Basic Prediction With TensorFlow

In this notebook, we will create a simple model to predict the *y* value of a given *x* value in a fixed relationship.

## Short Notes

This notebook is written on Google Colab, which can be accessed from [here](https://colab.research.google.com/drive/1sZcO590ZVAcaGLoUCHtNr7tLvsFABAp6?usp=sharing).

The problem is inspired by one of the Google Cloud Skills Boost's Hands-On Lab (GSP637 - TensorFlow: Qwik Start), which is linked [here](https://www.cloudskillsboost.google/focuses/7639?parent=catalog).

## Problem Description

The training data looks like this:

| *x* | -1 | 0 | 1 | 2 | 3 | 4 |
| --- | --- | --- | --- | --- | --- | --- |
| *y* | -2 | 1 | 4 | 7 | 10 | 13 |

From this table, we can clearly see that the relationship between *y* and *x* has the function of *y* = 3*x* + 1

## Import Necessary Packages

We will import both TensorFlow and Numpy modules.

For more information on installing the packages locally, check out [TensorFlow Installation Guide](https://www.tensorflow.org/install) and [Numpy Installation Guide](https://numpy.org/install/).

In [ ]:
import tensorflow as tf
import numpy as np

## Preparing Training Data

We will now create our numpy arrays storing the training data where the feature array is `xs` and label array is `ys`.

In [ ]:
xs = np.array([-1.0, 0.0, 1.0, 2.0, 3.0, 4.0], dtype=float)
ys = np.array([-2.0, 1.0, 4.0, 7.0, 10.0, 13.0], dtype=float)

## Model Creation

Notice that the function is one-to-one, meaning that each feature maps to exactly one label only. 

Since there is only 1 output per input, we will create a [Sequential](https://www.tensorflow.org/api_docs/python/tf/keras/Sequential) model. Due to the simplicity of the problem, we only need a model with 1 layer and 1 neuron.

In [ ]:
model = tf.keras.Sequential([
    tf.keras.layers.Dense(1, input_shape=(1,))  # Layer 1
])

The `input_shape` is `(1,)` because our feature array is a 1D array.

## Model Compilation

We will now compile our model with [SGD](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/SGD) Optimizer and [MSE](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/mean_squared_error) Loss Function. 

The model uses the Optimizer to make another guess. **Based on the Loss Function's result**, it will try to minimize the loss. Read more on other supported [Optimizers](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers).

The Loss Function measures the guessed answers against the known correct answers and **measures how well or how badly it did**. Read more on other supported [Loss Functions](https://www.tensorflow.org/api_docs/python/tf/keras/losses).

In [ ]:
model.compile(optimizer='sgd', loss='mse')

## Model Training

After the model completes its compilation, we will now feed in our training data to train our model.

The model takes in feature array `xs` and label array `ys`, with 500 training [epochs](https://www.baeldung.com/cs/epoch-neural-networks#epoch-in-neural-networks).

In [ ]:
model.fit(xs, ys, epochs=500)

Epoch 1/500
1/1 [==============================] - 1s 511ms/step - loss: 55.9806
Epoch 2/500
1/1 [==============================] - 0s 8ms/step - loss: 44.0430
Epoch 3/500
1/1 [==============================] - 0s 6ms/step - loss: 34.6510
Epoch 4/500
1/1 [==============================] - 0s 6ms/step - loss: 27.2619
Epoch 5/500
1/1 [==============================] - 0s 8ms/step - loss: 21.4484
Epoch 6/500
1/1 [==============================] - 0s 6ms/step - loss: 16.8747
Epoch 7/500
1/1 [==============================] - 0s 6ms/step - loss: 13.2763
Epoch 8/500
1/1 [==============================] - 0s 6ms/step - loss: 10.4453
Epoch 9/500
1/1 [==============================] - 0s 6ms/step - loss: 8.2180
Epoch 10/500
1/1 [==============================] - 0s 6ms/step - loss: 6.4656
Epoch 11/500
1/1 [==============================] - 0s 6ms/step - loss: 5.0869
Epoch 12/500
1/1 [==============================] - 0s 6ms/step - loss: 4.0022
Epoch 13/500
1/1 [==============================] -

## Model Prediction

After the model completes training, we can now make predictions using our model.

We will predict the *y* value when *x* is 10.0.

In [ ]:
prediction = model.predict([10.0])
print(prediction)

[[31.000422]]


In this occurence, our model predicted that value of *y* to be 31.000422 when *x* is 10.0. 

The reason that our model does not predict the value of *y* to be exactly 31.0 is because **neural networks deal with probabilities**. It calculated that there is a very high probability that the relationship between *x* and *y* is *y* = 3*x* + 1, but with only 6 data points it can't know for sure. As a result, the result for 10 is very close to 31, but not necessarily 31.